pulser-simulation/pulser_simulation/simresults.py

```diff
@@ -524,25 +524,34 @@ def sample_state(
         if self._meas_errors is None:
             return sampled_state
 
-        detected_sample_dict: Counter = Counter()
+        detected_sample_dict: defaultdict = defaultdict(int)
+        eps = self._meas_errors["epsilon"]
+        eps_p = self._meas_errors["epsilon_prime"]
+
+        if eps_p == 0.0 and eps == 0.0:
+            return sampled_state
+
         for shot, n_detects in sampled_state.items():
-            eps = self._meas_errors["epsilon"]
-            eps_p = self._meas_errors["epsilon_prime"]
             # Shot as an array of 1s and 0s
             shot_arr = np.array(list(shot), dtype=int)
+
             # Probability of flipping each bit
-            flip_probs = np.array([eps_p if x == "1" else eps for x in shot])
+            flip_probs = np.where(shot_arr == 1, eps_p, eps)
+
             # 1 if it flips, 0 if it stays the same
             flips = (
                 np.random.uniform(size=(n_detects, len(flip_probs)))
                 < flip_probs
-            ).astype(int)
+            )
+
             # XOR betwen the original array and the flips
             # Gives an array of n_detects individual shots
             new_shots = shot_arr ^ flips
```

Comment:
Note: I don't think we really care about the order of the lines in `shot_matrix`, it is perhaps easier to build a matrix of containing `n_detect` times the list of `shots`

Reply:
> I think this code can be vectorized. To me, what costs a lot is the multiple generation of a random matrix (one per instance in the for loop), and its comparison to compute `flips` and `new_shots`.
> 
> My idea is to create a `shot_matrix` that would be composed of `len(shot)*n_detect` lines and `len(shot)` columns (a matrix composed of `n_detect` times the array `shot` for each `shot` in `sampled_state.keys()`) `flip_probs` then becomes a matrix. `flips` is created by comparing a random matrix of size `(len(shot)*n_detect, len(shot))` with `flip_probs`. `new_shots` becomes a matrix and last for loop around `new_shots` is unchanged.
> 
> To perform this vectorization, it might be faster to use exclusively numpy functions (`np.bitwise_xor` is eventually faster than `^`, and combining `np.bincount` and `np.array2string` seems more relevant to build the final Counter than using `dectected_sample_dict`)

I can make a try, the performance will increase (not so much IMO): in addition will require more ram.  I don't understand the last part of `np.bincount` / `np.array2string` (btw ^ is a short-hand for `np.bitwise_xor`), can you explain further?

EDIT: I implemented it

```python
        shots = list(sampled_state.keys())
        n_detects_list = list(sampled_state.values())

        # Convert shots to a 2D array
        shot_arr = np.array([list(shot) for shot in shots], dtype=int)
        # Compute flip probabilities
        flip_probs = np.where(shot_arr == 1, eps_p, eps)
        # Repeat flip_probs based on n_detects_list
        flip_probs_repeated = np.repeat(flip_probs, n_detects_list, axis=0)
        # Generate random matrix of shape (sum(n_detects_list), len(shot))
        random_matrix = np.random.uniform(size=(np.sum(n_detects_list), len(shot_arr[0])))
        # Compare random matrix with flip probabilities
        flips = random_matrix < flip_probs_repeated
        # Perform XOR between original array and flips
        new_shots = shot_arr.repeat(n_detects_list, axis=0) ^ flips
        # Count all the new_shots
        detected_sample_dict = Counter(map(tuple, new_shots))
        
        return Counter(
            {"".join(map(str, k)): v for k, v in detected_sample_dict.items()}
        )
```

Performance are improved on increasing samples_per_run, but it increases also the RAM usage. 

Before:
```
Samples_per_run: 1	Elapsed: 3.863002061843872 seconds
Samples_per_run: 10	Elapsed: 4.330973863601685 seconds
Samples_per_run: 100	Elapsed: 6.834812164306641 seconds
Samples_per_run: 500	Elapsed: 14.479618549346924 seconds
Samples_per_run: 1000	Elapsed: 23.99378752708435 seconds
Samples_per_run: 2000	Elapsed: 42.217456579208374 seconds
Samples_per_run: 5000	Elapsed: 100.99924063682556 seconds
Samples_per_run: 10000	Elapsed: 203.1134431362152 seconds
```

After:
```
Samples_per_run: 1	Elapsed: 4.041079044342041 seconds
Samples_per_run: 10	Elapsed: 4.325520038604736 seconds
Samples_per_run: 100	Elapsed: 6.078380107879639 seconds
Samples_per_run: 500	Elapsed: 13.140138626098633 seconds
Samples_per_run: 1000	Elapsed: 21.684226989746094 seconds
Samples_per_run: 2000	Elapsed: 38.53966021537781 seconds
Samples_per_run: 5000	Elapsed: 90.3535590171814 seconds
Samples_per_run: 10000	Elapsed: 169.49966073036194 seconds
```

Reply:
Thanks for having a look to it ! Looks nice to me, what about you @HGSilveri ? I don't see any issue in using more RAM for a little speedup, but perhaps the tradeoff is not interesting.
Forget about the bincount/array2string thing, `Counter` is what you want ;)

Reply:
It's faster and pretty well documented, so I don't have objections at first glance.

Reply:
Yep, just pushed!